File: frontend/struct/src/functions/sortings/shellSort.ts
import { SortStep } from "@/interfaces/interface";

export const shellSortStep = (
  arr: number[],
  step: SortStep,
  ascending: boolean = true,
): { sortedArray: number[]; step: SortStep } => {
  const sortedArray = [...arr];
  const n = sortedArray.length;

  let gap = step.gap !== undefined ? step.gap : Math.floor(n / 2);
  let { i, j } = step;

  if (gap <= 0) {
    return { sortedArray, step: { i: n, j: n, gap: 0 } };
  }

  if (i === undefined || i >= n) {
    i = gap;
  }

  while (i < n) {
    const temp = sortedArray[i];
    j = i;

    if (ascending) {
      while (j >= gap && sortedArray[j - gap] > temp) {
        sortedArray[j] = sortedArray[j - gap];
        j -= gap;
      }
    } else {
      while (j >= gap && sortedArray[j - gap] < temp) {
        sortedArray[j] = sortedArray[j - gap];
        j -= gap;
      }
    }

    sortedArray[j] = temp;
    i++;
  }

  gap = Math.floor(gap / 2);
  i = gap;

  return { sortedArray, step: { i, j, gap } };
};

File: src/shell.sort.cpp
template <typename T> void shellSort(T *arr, int n, bool ascending = true) {
  for (int gap = n / 2; gap > 0; gap /= 2) {
    for (int i = gap; i < n; ++i) {
      T   temp = arr[i];
      int j;

      if (ascending) {
        for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
          arr[j] = arr[j - gap];
        }
      } else {
        for (j = i; j >= gap && arr[j - gap] < temp; j -= gap) {
          arr[j] = arr[j - gap];
        }
      }

      arr[j] = temp;
    }
  }
}
